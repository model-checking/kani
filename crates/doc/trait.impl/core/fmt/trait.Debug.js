(function() {
    var implementors = Object.fromEntries([["kani",[["impl <a class=\"trait\" href=\"https://doc.rust-lang.org/nightly/core/fmt/trait.Debug.html\" title=\"trait core::fmt::Debug\">Debug</a> for <a class=\"enum\" href=\"kani/enum.AllocationStatus.html\" title=\"enum kani::AllocationStatus\">AllocationStatus</a>"],["impl&lt;'a, T: <a class=\"trait\" href=\"https://doc.rust-lang.org/nightly/core/fmt/trait.Debug.html\" title=\"trait core::fmt::Debug\">Debug</a>&gt; <a class=\"trait\" href=\"https://doc.rust-lang.org/nightly/core/fmt/trait.Debug.html\" title=\"trait core::fmt::Debug\">Debug</a> for <a class=\"struct\" href=\"kani/struct.ArbitraryPointer.html\" title=\"struct kani::ArbitraryPointer\">ArbitraryPointer</a>&lt;'a, T&gt;"],["impl&lt;const BYTES: <a class=\"primitive\" href=\"https://doc.rust-lang.org/nightly/std/primitive.usize.html\">usize</a>&gt; <a class=\"trait\" href=\"https://doc.rust-lang.org/nightly/core/fmt/trait.Debug.html\" title=\"trait core::fmt::Debug\">Debug</a> for <a class=\"struct\" href=\"kani/struct.PointerGenerator.html\" title=\"struct kani::PointerGenerator\">PointerGenerator</a>&lt;BYTES&gt;"]]]]);
    if (window.register_implementors) {
        window.register_implementors(implementors);
    } else {
        window.pending_implementors = implementors;
    }
})()
//{"start":57,"fragment_lengths":[1100]}