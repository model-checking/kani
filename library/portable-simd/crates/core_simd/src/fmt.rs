use crate::simd::{LaneCount, Simd, SimdElement, SupportedLaneCount};
use core::fmt;

macro_rules! impl_fmt_trait {
    { $($trait:ident,)* } => {
        $(
            impl<T, const LANES: usize> fmt::$trait for Simd<T, LANES>
            where
                LaneCount<LANES>: SupportedLaneCount,
                T: SimdElement + fmt::$trait,
            {
                fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                    #[repr(transparent)]
                    struct Wrapper<'a, T: fmt::$trait>(&'a T);

                    impl<T: fmt::$trait> fmt::Debug for Wrapper<'_, T> {
                        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                            self.0.fmt(f)
                        }
                    }

                    f.debug_list()
                        .entries(self.as_array().iter().map(|x| Wrapper(x)))
                        .finish()
                }
            }
        )*
    }
}

impl_fmt_trait! {
    Debug,
    Binary,
    LowerExp,
    UpperExp,
    Octal,
    LowerHex,
    UpperHex,
}
