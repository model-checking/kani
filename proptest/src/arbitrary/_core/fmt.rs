//-
// Copyright 2017, 2018 The proptest developers
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

// SPDX-License-Identifier: Apache-2.0 OR MIT
//
// Modifications Copyright Kani Contributors
// See GitHub history for details.

//! Arbitrary implementations for `std::fmt`.

use core::fmt::Error;
arbitrary!(Error; Error);

#[cfg(all(test, not(kani)))]
mod test {
    no_panic_test!(error => Error);
}
