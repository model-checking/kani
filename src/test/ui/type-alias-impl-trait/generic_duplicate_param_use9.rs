#![feature(type_alias_impl_trait)]

use std::fmt::Debug;

fn main() {}

type Two<A, B> = impl Debug;
//~^ ERROR the trait bound `A: Foo` is not satisfied in `(A, B, <A as Foo>::Bar)`
//~| ERROR `A` doesn't implement `Debug`
//~| ERROR `B` doesn't implement `Debug`

trait Foo {
    type Bar: Debug;
    const BAR: Self::Bar;
}

fn two<T: Debug + Foo, U: Debug>(t: T, u: U) -> Two<T, U> {
    (t, u, T::BAR)
}

fn three<T: Debug, U: Debug>(t: T, u: U) -> Two<T, U> {
    (t, u, 42) //~^ ERROR concrete type differs from previous
}
