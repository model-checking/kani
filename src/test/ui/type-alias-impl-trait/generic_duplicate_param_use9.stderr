error: concrete type differs from previous defining opaque type use
  --> $DIR/generic_duplicate_param_use9.rs:21:1
   |
LL | fn three<T: Debug, U: Debug>(t: T, u: U) -> Two<T, U> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `(A, B, <A as Foo>::Bar)`, got `(A, B, i32)`
   |
note: previous use here
  --> $DIR/generic_duplicate_param_use9.rs:17:1
   |
LL | fn two<T: Debug + Foo, U: Debug>(t: T, u: U) -> Two<T, U> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `A: Foo` is not satisfied in `(A, B, <A as Foo>::Bar)`
  --> $DIR/generic_duplicate_param_use9.rs:7:18
   |
LL | type Two<A, B> = impl Debug;
   |                  ^^^^^^^^^^ within `(A, B, <A as Foo>::Bar)`, the trait `Foo` is not implemented for `A`
   |
   = note: required because it appears within the type `(A, B, <A as Foo>::Bar)`
help: consider restricting type parameter `A`
   |
LL | type Two<A: Foo, B> = impl Debug;
   |           +++++

error[E0277]: `A` doesn't implement `Debug`
  --> $DIR/generic_duplicate_param_use9.rs:7:18
   |
LL | type Two<A, B> = impl Debug;
   |                  ^^^^^^^^^^ `A` cannot be formatted using `{:?}` because it doesn't implement `Debug`
   |
   = note: required because of the requirements on the impl of `Debug` for `(A, B, <A as Foo>::Bar)`
help: consider restricting type parameter `A`
   |
LL | type Two<A: std::fmt::Debug, B> = impl Debug;
   |           +++++++++++++++++

error[E0277]: `B` doesn't implement `Debug`
  --> $DIR/generic_duplicate_param_use9.rs:7:18
   |
LL | type Two<A, B> = impl Debug;
   |                  ^^^^^^^^^^ `B` cannot be formatted using `{:?}` because it doesn't implement `Debug`
   |
   = note: required because of the requirements on the impl of `Debug` for `(A, B, <A as Foo>::Bar)`
help: consider restricting type parameter `B`
   |
LL | type Two<A, B: std::fmt::Debug> = impl Debug;
   |              +++++++++++++++++

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0277`.
